lib/Sema/CSRanking.cpp

```diff
@@ -936,16 +937,29 @@ SolutionCompareResult ConstraintSystem::compareSolutions(
         choice2.getKind() == OverloadChoiceKind::DeclViaDynamic;
 
     // Determine whether one declaration is more specialized than the other.
+    //
+    // If solving for code completion, we don't penalize the less specialized
+    // overload, as completions generated from it once inserted may result in
+    // it being the only viable overload:
+    //
+    //  func foo(_ x: Int) -> Int {}
+    //  func foo<T>(_ x: T) -> String {}
+    //  foo(3).<complete here> // Still want solutions with for both foo
+    //                         // overloads - String and Int members are both
+    //                         // valid here.
+    //
     bool firstAsSpecializedAs = false;
     bool secondAsSpecializedAs = false;
     if (isDeclAsSpecializedAs(cs.DC, decl1, decl2,
                               isDynamicOverloadComparison)) {
-      score1 += weight;
+      if (!isForCodeCompletion)
```

Comment:
Would "less specialize" in this case always mean generic vs. non-generic overloads? `isDeclAsSpecializedAs` does more checks than that (i.e. it also checks whether there is subtyping relationship between generic arguments), maybe this flag should be moved there to prevent some of the undesirable checks, or should code completion mode only filter based on scores and avoid checking specialization all together? I think ultimately the problem is going to be the same as with previous check which circumvented filtering all together - too many partial solutions slowing down inference...